# 🎭 Theaters & Cinemas in Berlin — integration of a new data layer on theaters into the database.


This issue outlines the integration of a new data layer on Theaters in Berlin into the database.

This work is part of EPIC 2: Data Foundation & Frontend Context, which focuses on building the data layers for the MVP.

✅ CSV saved to: /Users/mariannagokova/Documents/DA_PROJECT/Thetres_Berlin_Task/clear_ready_git/source/theaters_berlin_db_ready.csv


✅ GeoJSON saved to: /Users/mariannagokova/Documents/DA_PROJECT/Thetres_Berlin_Task/clear_ready_git/source/theaters_berlin_db_ready.geojson


📦 CSV size: 77.8 KB
📦 GeoJSON size: 226.9 KB

🧾 Saved dataset summary:
Rows: 282
Columns: 26
CRS: epsg:4326

In [ ]:
"""
CREATE TABLE theaters (
    theater_id         VARCHAR(64) PRIMARY KEY,         -- Unique stable ID (e.g., hash or UUID)
    name               VARCHAR(255) NOT NULL,           -- Official name of the theatre or cinema
    name_key           VARCHAR(255),                    -- Normalized lowercase name (slug, key)
    place_type         VARCHAR(50),                     -- Type: "cinema", "theatre", etc.
    operator           VARCHAR(255),                    -- Organization or company operating it
    opening_hours      VARCHAR(255),                    -- OSM-style hours string
    wheelchair         VARCHAR(50),                     -- Accessibility info: "yes", "no", "limited"
    screen             INTEGER,                         -- Number of screens (for cinemas)
    website            VARCHAR(255),                    -- Official website URL
    phone              VARCHAR(100),                    -- Contact phone number
    email              VARCHAR(255),                    -- Contact email address
    addr_full          VARCHAR(255),                    -- Full formatted address
    addr_street        VARCHAR(255),                    -- Street name
    addr_housenumber   VARCHAR(50),                     -- House or building number
    addr_postcode      VARCHAR(20),                     -- Postal code
    addr_city          VARCHAR(100),                    -- City (usually "Berlin")
    addr_country       VARCHAR(100),                    -- Country (usually "Germany")
    theatre_tags       TEXT,                            -- Raw tags or classification info from OSM/Wikidata
    theatre_category   VARCHAR(100),                    -- Derived label: e.g., "performing arts", "independent cinema"
    district_id        VARCHAR(10),                     -- LOR district code
    district           VARCHAR(100),                    -- LOR district name
    neighborhood_id    VARCHAR(10),                     -- LOR neighborhood (Ortsteil) code
    longitude          DECIMAL(9,6),                    -- WGS84 coordinate (lon)
    latitude           DECIMAL(9,6),                    -- WGS84 coordinate (lat)
    last_updated       TIMESTAMP DEFAULT CURRENT_TIMESTAMP  -- Timestamp when data last updated
    CONSTRAINT district_id_fk FOREIGN KEY (district_id)
        REFERENCES berlin_data.districts(district_id)
        ON DELETE RESTRICT
        ON UPDATE CASCADE
);
"""

Make sure Python environment has psycopg2 or SQLAlchemy installed:

In [ ]:
#pip install psycopg2-binary sqlalchemy pandas


In [13]:
import psycopg2
import pandas as pd

conn = psycopg2.connect(
    host="127.0.0.1",
    port=5433,                     # tunnel port
    user="marianna_gokova",              # <-- DB user
    password="6n2b8nw9IfmNdyYY",      # <-- DB password
    dbname="layereddb",         # <-- database name
    sslmode="require"              # use "verify-full" if RDS enforces cert validation
)

# test the connection
with conn.cursor() as cur:
    cur.execute("SELECT current_database(), current_user;")
    print(cur.fetchall())


[('layereddb', 'marianna_gokova')]


In [14]:
# list schemas
query = """
SELECT schema_name
FROM information_schema.schemata
ORDER BY schema_name;
"""
schemas = pd.read_sql(query, conn)
schemas

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/385576128.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  schemas = pd.read_sql(query, conn)


schema_name
0       berlin_labels
1  berlin_recommender
2  berlin_source_data
3      dashboard_data
4  information_schema
5          pg_catalog
6              public

In [15]:
#See tables per schema
import pandas as pd

def show_tables(schema):
    return pd.read_sql(f"""
        SELECT table_schema, table_name
        FROM information_schema.tables
        WHERE table_schema = '{schema}'
        ORDER BY table_name;
    """, conn)

show_tables('berlin_labels')



/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/1795356164.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(f"""


table_schema              table_name
0  berlin_labels     district_attributes
1  berlin_labels  district_features_test
2  berlin_labels         district_labels
3  berlin_labels     district_labels_new
4  berlin_labels     neighborhood_labels

In [16]:
#Inspect the source districts table (just to be sure)

import pandas as pd
pd.read_sql("""
SELECT column_name, data_type
FROM information_schema.columns
WHERE table_schema='berlin_source_data' AND table_name='districts'
ORDER BY ordinal_position;
""", conn)

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/231728245.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql("""


column_name          data_type
0  district_id  character varying
1     district  character varying
2     geometry       USER-DEFINED

In [17]:
show_tables('berlin_source_data')

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/1795356164.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(f"""


table_schema                        table_name
0   berlin_source_data                             banks
1   berlin_source_data                        bike_lanes
2   berlin_source_data                    bus_tram_stops
3   berlin_source_data                  crime_statistics
4   berlin_source_data                    dental_offices
5   berlin_source_data          district_attributes_test
6   berlin_source_data         district_level_aggregated
7   berlin_source_data                         districts
8   berlin_source_data                districts_pop_stat
9   berlin_source_data                              gyms
10  berlin_source_data                         hospitals
11  berlin_source_data              hospitals_refactored
12  berlin_source_data                     kindergartens
13  berlin_source_data                       land_prices
14  berlin_source_data                long_term_listings
15  berlin_source_data                             malls
16  berlin_source_data     milieuschutz_protection_zones
17  berlin_source_data                     neighborhoods
18  berlin_source_data                             parks
19  berlin_source_data                        pharmacies
20  berlin_source_data                       playgrounds
21  berlin_source_data                             pools
22  berlin_source_data                  pools_refactored
23  berlin_source_data                      post_offices
24  berlin_source_data               regional_statistics
25  berlin_source_data       rent_stats_per_neighborhood
26  berlin_source_data                             sbahn
27  berlin_source_data                           schools
28  berlin_source_data               short_term_listings
29  berlin_source_data           social_clubs_activities
30  berlin_source_data                      supermarkets
31  berlin_source_data          test_table_george_smelin
32  berlin_source_data                          theaters
33  berlin_source_data                             ubahn
34  berlin_source_data                      universities
35  berlin_source_data                            venues
36  berlin_source_data                veterinary_clinics
37  berlin_source_data  veterinary_clinics_martin_svitek

In [18]:
with conn, conn.cursor() as cur:
    cur.execute("""
    
CREATE TABLE IF NOT EXISTS berlin_source_data.theaters (
    theater_id         VARCHAR(64) PRIMARY KEY,         -- Unique stable ID (e.g., hash or UUID)
    name               VARCHAR(255) NOT NULL,           -- Official name of the theatre or cinema
    name_key           VARCHAR(255),                    -- Normalized lowercase name (slug, key)
    place_type         VARCHAR(50),                     -- Type: "cinema", "theatre", etc.
    operator           VARCHAR(255),                    -- Organization or company operating it
    opening_hours      VARCHAR(255),                    -- OSM-style hours string
    wheelchair         VARCHAR(50),                     -- Accessibility info: "yes", "no", "limited"
    screen             INTEGER,                         -- Number of screens (for cinemas)
    website            VARCHAR(255),                    -- Official website URL
    phone              VARCHAR(100),                    -- Contact phone number
    email              VARCHAR(255),                    -- Contact email address
    addr_full          VARCHAR(255),                    -- Full formatted address
    addr_street        VARCHAR(255),                    -- Street name
    addr_housenumber   VARCHAR(50),                     -- House or building number
    addr_postcode      VARCHAR(20),                     -- Postal code
    addr_city          VARCHAR(100),                    -- City (usually "Berlin")
    addr_country       VARCHAR(100),                    -- Country (usually "Germany")
    theatre_tags       TEXT,                            -- Raw tags or classification info from OSM/Wikidata
    theatre_category   VARCHAR(100),                    -- Derived label: e.g., "performing arts", "independent cinema"
    district_id        VARCHAR(10),                     -- LOR district code
    district           VARCHAR(100),                    -- LOR district name
    neighborhood_id    VARCHAR(10),                     -- LOR neighborhood (Ortsteil) code
    longitude          DECIMAL(9,6),                    -- WGS84 coordinate (lon)
    latitude           DECIMAL(9,6),                    -- WGS84 coordinate (lat)
    last_updated       TIMESTAMP DEFAULT CURRENT_TIMESTAMP,  -- Timestamp when data last updated
    CONSTRAINT district_id_fk FOREIGN KEY (district_id)
        REFERENCES berlin_source_data.districts(district_id)
        ON DELETE RESTRICT
        ON UPDATE CASCADE
);
    """)



In [19]:
#Quick check:
import pandas as pd

pd.read_sql("""
SELECT table_schema, table_name
FROM information_schema.tables
WHERE table_schema='berlin_source_data' AND table_name='theaters';
""", conn)



/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/3426442766.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql("""


table_schema table_name
0  berlin_source_data   theaters

In [25]:

import pandas as pd
from sqlalchemy import create_engine

# 1) Read CSV and fix column types for screen-Integer, not float
df = pd.read_csv("/Users/mariannagokova/Documents/DA_PROJECT/Thetres_Berlin_Task/clear_ready_git/source/theaters_berlin_db_ready.csv")

# If "screen" exists, convert to integer safely
if "screen" in df.columns:
    df["screen"] = pd.to_numeric(df["screen"], errors="coerce").fillna(0).astype(int)

# 2) Connect via SQLAlchemy (using your existing tunnel)
engine = create_engine("postgresql+psycopg2://marianna_gokova:6n2b8nw9IfmNdyYY@127.0.0.1:5433/layereddb?sslmode=require")

# 3) Load directly
df.to_sql(
    "theaters",
    schema="berlin_source_data",
    con=engine,
    if_exists="append",  # add data to existing table
    index=False
)

# 4) Check result
pd.read_sql("SELECT COUNT(*) FROM berlin_source_data.theaters;", engine)


IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "theaters_pkey"
DETAIL:  Key (theater_id)=(thb_1fed95705f16) already exists.

[SQL: INSERT INTO berlin_source_data.theaters (theater_id, name, name_key, place_type, operator, opening_hours, wheelchair, screen, website, phone, email, addr_full, addr_street, addr_housenumber, addr_postcode, addr_city, addr_country, theatre_tags, theat ... 145011 characters truncated ... rict__281)s, %(neighborhood_id__281)s, %(longitude__281)s, %(latitude__281)s, %(last_updated__281)s)]
[parameters: {'operator__0': 'Filmrausch Moabit e.V.', 'theatre_tags__0': None, 'name__0': 'Filmrauschpalast', 'addr_full__0': 'Lehrter Straße 35 10557 Berlin', 'latitude__0': 52.534378399999994, 'addr_country__0': 'DE', 'district__0': 'Mitte', 'addr_housenumber__0': '35', 'district_id__0': 11001001, 'email__0': 'info@filmrausch.de', 'last_updated__0': '2025-10-12T23:39:15+02:00', 'addr_city__0': 'Berlin', 'longitude__0': 13.3596231, 'screen__0': 1, 'wheelchair__0': 'no', 'theatre_category__0': None, 'website__0': 'https://www.filmrausch.de/', 'opening_hours__0': None, 'addr_street__0': 'Lehrter Straße', 'name_key__0': 'filmrauschpalast', 'place_type__0': 'cinema', 'theater_id__0': 'thb_1fed95705f16', 'phone__0': '+49303844344', 'neighborhood_id__0': 102, 'addr_postcode__0': 10557.0, 'operator__1': None, 'theatre_tags__1': None, 'name__1': 'Friedrichstadt-Palast', 'addr_full__1': 'Friedrichstraße 107 10117 Berlin', 'latitude__1': 52.5239216, 'addr_country__1': 'DE', 'district__1': 'Mitte', 'addr_housenumber__1': '107', 'district_id__1': 11001001, 'email__1': None, 'last_updated__1': '2025-10-12T23:39:15+02:00', 'addr_city__1': 'Berlin', 'longitude__1': 13.3888786, 'screen__1': 0, 'wheelchair__1': 'yes', 'theatre_category__1': None, 'website__1': 'https://www.palast.berlin/', 'opening_hours__1': None, 'addr_street__1': 'Friedrichstraße', 'name_key__1': 'friedrichstadt-palast', 'place_type__1': 'theatre', 'theater_id__1': 'thb_957cf1213125', 'phone__1': '+493023262326', 'neighborhood_id__1': 101, 'addr_postcode__1': 10117.0 ... 6950 parameters truncated ... 'operator__280': 'Max Mohr', 'theatre_tags__280': 'circus; variety', 'name__280': 'Zirkus Mond', 'addr_full__280': 'Lilli-Henoch-Straße Berlin', 'latitude__280': 52.542281641466914, 'addr_country__280': 'DE', 'district__280': 'Pankow', 'addr_housenumber__280': None, 'district_id__280': 11003003, 'email__280': 'zirkusmond@posteo.de', 'last_updated__280': '2025-10-12T23:39:15+02:00', 'addr_city__280': 'Berlin', 'longitude__280': 13.432500433400511, 'screen__280': 0, 'wheelchair__280': None, 'theatre_category__280': 'circus', 'website__280': 'https://zirkusmond.de/', 'opening_hours__280': None, 'addr_street__280': 'Lilli-Henoch-Straße', 'name_key__280': 'zirkus-mond', 'place_type__280': 'theatre', 'theater_id__280': 'thb_cbcf8e2e7421', 'phone__280': '+4915770280810', 'neighborhood_id__280': 301, 'addr_postcode__280': None, 'operator__281': 'Shakespeare Company Berlin e.V.', 'theatre_tags__281': 'drama', 'name__281': 'Shakespeare Company Berlin', 'addr_full__281': 'Munsterdamm 80 12169 Berlin', 'latitude__281': 52.45650912442611, 'addr_country__281': 'DE', 'district__281': 'Steglitz-Zehlendorf', 'addr_housenumber__281': '80', 'district_id__281': 11006006, 'email__281': 'info@shakespeare-company.de', 'last_updated__281': '2025-10-12T23:39:15+02:00', 'addr_city__281': 'Berlin', 'longitude__281': 13.34983712450778, 'screen__281': 0, 'wheelchair__281': None, 'theatre_category__281': 'drama', 'website__281': 'https://shakespeare-company.de/', 'opening_hours__281': None, 'addr_street__281': 'Munsterdamm', 'name_key__281': 'shakespeare-company-berlin', 'place_type__281': 'theatre', 'theater_id__281': 'thb_a99ca9e3b894', 'phone__281': None, 'neighborhood_id__281': 601, 'addr_postcode__281': 12169.0}]
(Background on this error at: https://sqlalche.me/e/20/gkpj)

# 🧩 Sanity check block

In [26]:
import pandas as pd

# ✅ 1. Count total rows
total = pd.read_sql("""
SELECT COUNT(*) AS total_rows
FROM berlin_source_data.theaters;
""", engine)

# ✅ 2. Check for invalid or missing district_id (FK consistency)
missing_fk = pd.read_sql("""
SELECT COUNT(*) AS missing_districts
FROM berlin_source_data.theaters t
LEFT JOIN berlin_source_data.districts d
  ON d.district_id = t.district_id
WHERE d.district_id IS NULL OR t.district_id IS NULL;
""", engine)

# ✅ 3. Check for duplicate theater_id (should be unique)
dups_id = pd.read_sql("""
SELECT COUNT(*) AS duplicate_ids
FROM (
  SELECT theater_id
  FROM berlin_source_data.theaters
  GROUP BY theater_id
  HAVING COUNT(*) > 1
) dup;
""", engine)

# ✅ 4. Check for potential duplicates by name + address
dups_name = pd.read_sql("""
SELECT COUNT(*) AS duplicate_name_address
FROM (
  SELECT name, addr_street, addr_housenumber
  FROM berlin_source_data.theaters
  GROUP BY name, addr_street, addr_housenumber
  HAVING COUNT(*) > 1
) dup;
""", engine)

# ✅ 5. Distribution by district
by_district = pd.read_sql("""
SELECT district_id, COUNT(*) AS num_theaters
FROM berlin_source_data.theaters
GROUP BY district_id
ORDER BY num_theaters DESC;
""", engine)

# ✅ 6. Distribution by place_type
by_type = pd.read_sql("""
SELECT place_type, COUNT(*) AS num
FROM berlin_source_data.theaters
GROUP BY place_type
ORDER BY num DESC;
""", engine)

# ✅ 7. Quick geographic sanity check (any missing coordinates?)
geo_check = pd.read_sql("""
SELECT
  COUNT(*) FILTER (WHERE longitude IS NULL OR latitude IS NULL) AS missing_coordinates,
  COUNT(*) FILTER (WHERE longitude < 5 OR longitude > 20 OR latitude < 47 OR latitude > 56) AS out_of_bounds
FROM berlin_source_data.theaters;
""", engine)

# 🔍 Display results
print("===== Sanity Check Summary =====")
display(total)
display(missing_fk)
display(dups_id)
display(dups_name)
display(geo_check)

print("===== Distribution by District =====")
display(by_district.head(10))

print("===== Distribution by Place Type =====")
display(by_type.head(10))


===== Sanity Check Summary =====


total_rows
0         282

missing_districts
0                  0

duplicate_ids
0              0

duplicate_name_address
0                       6

missing_coordinates  out_of_bounds
0                    0              0

===== Distribution by District =====


district_id  num_theaters
0    11001001            63
1    11002002            44
2    11004004            38
3    11003003            29
4    11007007            22
5    11009009            22
6    11008008            20
7    11006006            15
8    11010010             9
9    11011011             8

===== Distribution by Place Type =====


place_type  num
0    theatre  190
1     cinema   92

### 🧩 Step: Update `berlin_source_data.theaters` with Neighborhood Names

In this step, we enrich the existing database table **`berlin_source_data.theaters`** by adding neighborhood names based on the `neighborhood_id` reference.

The source file —  
`source/theatres_berlin_enriched_district.csv` —  
contains mappings between each `neighborhood_id` (Ortsteil code) and its corresponding `neighborhood` name.

To safely update the database:

1. The script connects to the PostgreSQL database `layereddb`.
2. It ensures the column **`neighborhood`** exists in the target table.
3. A **temporary table** (`tmp_theaters_enriched`) is created within the same session.  
   - Columns:  
     - `neighborhood_id VARCHAR(10)`  
     - `neighborhood VARCHAR(100)`  
4. The data from the CSV is bulk-inserted into this temporary table.
5. The main table `berlin_source_data.theaters` is updated:
   - `neighborhood` is filled or refreshed wherever the `neighborhood_id` matches.
   - Only rows with changed or missing neighborhood names are updated.
6. The temporary table is dropped automatically at the end of the session — no staging data remains.

**Result:**  
Each theater record now contains both `neighborhood_id` and its corresponding `neighborhood` name, ensuring complete LOR geographic enrichment in the Berlin dataset.


In [27]:
from sqlalchemy import create_engine, text
import pandas as pd
# --- Connection config ---
USER = "marianna_gokova"
PASSWORD = "6n2b8nw9IfmNdyYY"
HOST = "127.0.0.1"
PORT = 5433
DBNAME = "layereddb"

# --- Create SQLAlchemy engine ---
engine = create_engine(
    f"postgresql+psycopg2://{USER}:{PASSWORD}@{HOST}:{PORT}/{DBNAME}",
    connect_args={"sslmode": "require"}
)

# --- Query to detect missing or invalid neighborhood_id ---
query = text("""
SELECT 
    theater_id,
    name,
    district_id,
    neighborhood_id,
    neighborhood
FROM berlin_source_data.theaters
WHERE neighborhood_id IS NULL
   OR neighborhood_id = ''
   OR neighborhood_id NOT SIMILAR TO '^[0-9]{3,4}$'
ORDER BY district_id;
""")

# --- Execute and load to DataFrame ---
with engine.connect() as conn:
    df_missing_neigh = pd.read_sql(query, conn)

# --- Show summary ---
print(f"🧾 Total rows checked: {len(df_missing_neigh)}")
display(df_missing_neigh.head(10))

🧾 Total rows checked: 282


theater_id                  name district_id neighborhood_id  \
0  thb_d7b684833f96  Videokino Peliculoso    11001001             101   
1  thb_ab5cf1d9dfba  Komische Oper Berlin    11001001             101   
2  thb_e0da8d4b4612   Quatsch Comedy Club    11001001             101   
3  thb_4a7c4260ca84     Theater Mirakulum    11001001             101   
4  thb_e3cbde4f644e          Zeughauskino    11001001             101   
5  thb_5359071622cc   Hackesche Höfe Kino    11001001             101   
6  thb_47b251baf1dc          Kino Central    11001001             101   
7  thb_500c97d5d7e9     Deutsches Theater    11001001             101   
8  thb_8c996cc0e14c     Atze Musiktheater    11001001             105   
9  thb_2b46a472712d    Kino International    11001001             101   

  neighborhood  
0        Mitte  
1        Mitte  
2        Mitte  
3        Mitte  
4        Mitte  
5        Mitte  
6        Mitte  
7        Mitte  
8      Wedding  
9        Mitte

In [28]:
# --- Query: detect 4-digit neighborhood_id ---
query = text("""
SELECT 
    theater_id,
    name,
    district_id,
    neighborhood_id,
    neighborhood
FROM berlin_source_data.theaters
WHERE neighborhood_id ~ '^[0-9]{4}$'
ORDER BY district_id, neighborhood_id;
""")

# --- Execute and load to DataFrame ---
with engine.connect() as conn:
    df_4digit = pd.read_sql(query, conn)

# --- Summary ---
print(f"🏙️ Theatres with 4-digit neighborhood_id: {len(df_4digit)}")
display(df_4digit.head(22))

🏙️ Theatres with 4-digit neighborhood_id: 22


theater_id                        name district_id neighborhood_id  \
0   thb_b4ea6a9ef6ed                    Cabuwazi    11010010            1001   
1   thb_6d4b42916d6d  Berliner Tschechow Theater    11010010            1001   
2   thb_6f16930bbbbe             UCI Am Eastgate    11010010            1001   
3   thb_2b8dd6c9c9d3                       Arena    11010010            1001   
4   thb_218909bb7c4b             Theater am Park    11010010            1002   
5   thb_6a30166f72a9        Biesdorfer Parkbühne    11010010            1002   
6   thb_545241c8cf31                 Grüne Bühne    11010010            1005   
7   thb_a452daa99443                    CineStar    11010010            1005   
8   thb_fdaaad4c8ce1                   Die Kiste    11010010            1005   
9   thb_af0e1b19417c              Freilichtbühne    11011011            1101   
10  thb_3c254f253c43                Konzertbühne    11011011            1101   
11  thb_53838f24d15a              Freilichtbühne    11011011            1103   
12  thb_6972d9201f65      Theater an der Parkaue    11011011            1103   
13  thb_a6b918aace8e                  CineMotion    11011011            1109   
14  thb_25d0c47e8c72           Das Weite Theater    11011011            1111   
15  thb_948b8d0cf145    Theater Strahl, Ostkreuz    11011011            1112   
16  thb_689df4659585               Max-Taut-Aula    11011011            1112   
17  thb_866953d32b36                    CineStar    11012012            1202   
18  thb_2c82741d9d2c        Kulturversteck Tegel    11012012            1202   
19  thb_cb942bc2df22       Madi – Zelt der Sinne    11012012            1202   
20  thb_2604b3991aa1      Papiertheater Invisius    11012012            1205   
21  thb_f4118567ad9b           Ernst-Reuter-Saal    11012012            1209   

            neighborhood  
0                Marzahn  
1                Marzahn  
2                Marzahn  
3                Marzahn  
4               Biesdorf  
5               Biesdorf  
6            Hellersdorf  
7            Hellersdorf  
8            Hellersdorf  
9        Friedrichsfelde  
10       Friedrichsfelde  
11           Lichtenberg  
12           Lichtenberg  
13  Neu-Hohenschönhausen  
14             Fennpfuhl  
15           Rummelsburg  
16           Rummelsburg  
17                 Tegel  
18                 Tegel  
19                 Tegel  
20               Frohnau  
21              Wittenau

In [ ]:
import pandas as pd
import psycopg2
from psycopg2.extras import execute_values

CSV_PATH = "/Users/mariannagokova/Documents/DA_PROJECT/Thetres_Berlin_Task/clear_ready_git/source/theatres_berlin_enriched_district.csv"

conn = psycopg2.connect(
    host="127.0.0.1",
    port=5433,
    user="marianna_gokova",
    password="6n2b8nw9IfmNdyYY",
    dbname="layereddb",
    sslmode="require"
)

# --- Load dictionary (neighborhood_id, neighborhood) again (id = 4 digits) ---
df = pd.read_csv(
    CSV_PATH,
    dtype={"neighborhood_id": "string", "neighborhood": "string"},
    encoding="utf-8-sig"
)
df["neighborhood_id"] = df["neighborhood_id"].astype("string").str.strip().str.replace(r"\D","",regex=True).str.zfill(4)
df["neighborhood"]    = df["neighborhood"].astype("string").str.strip()
df = df.dropna(subset=["neighborhood_id"])
df = df[df["neighborhood"].fillna("").str.len() > 0]
df = df.drop_duplicates(subset=["neighborhood_id"])

try:
    with conn, conn.cursor() as cur:
        # 0) Backup (one time)
        cur.execute("""
            CREATE TABLE IF NOT EXISTS berlin_source_data.theaters_backup_neigh_final AS
            SELECT * FROM berlin_source_data.theaters;
        """)
        print("💾 Backup OK")

        # 1) Create/refresh temp dictionary table
        cur.execute("""
            CREATE TEMP TABLE IF NOT EXISTS tmp_neigh_dict (
                neighborhood_id CHAR(4) PRIMARY KEY,
                neighborhood    VARCHAR(100)
            ) ON COMMIT DROP;
            TRUNCATE tmp_neigh_dict;
        """)
        execute_values(
            cur,
            "INSERT INTO tmp_neigh_dict (neighborhood_id, neighborhood) VALUES %s;",
            list(df[["neighborhood_id","neighborhood"]].itertuples(index=False, name=None))
        )
        print(f"📚 Dictionary ready: {len(df)} ids")

        # 2) Normalize theaters.neighborhood_id to exactly 4 digits (digits only, LPAD)
        cur.execute("""
            UPDATE berlin_source_data.theaters
            SET neighborhood_id = LPAD(REGEXP_REPLACE(COALESCE(neighborhood_id,''), '\\D', '', 'g'), 4, '0')
            WHERE neighborhood_id IS NOT NULL
              AND (neighborhood_id !~ '^[0-9]{4}$' OR neighborhood_id LIKE '%[^0-9]%' );
        """)
        print(f"🔧 Normalized id length: {cur.rowcount} rows touched")

        # 3) Treat '0000' as unknown → NULL (easier to find/fix)
        cur.execute("""
            UPDATE berlin_source_data.theaters
            SET neighborhood_id = NULL
            WHERE neighborhood_id = '0000';
        """)
        print(f"🧽 '0000' -> NULL: {cur.rowcount} rows")

        # 4) Update neighborhood name from dictionary by id ( step, re-run to be sure)
        cur.execute("""
            UPDATE berlin_source_data.theaters t
            SET neighborhood = d.neighborhood
            FROM tmp_neigh_dict d
            WHERE t.neighborhood_id = d.neighborhood_id
              AND t.neighborhood IS DISTINCT FROM d.neighborhood;
        """)
        print(f"🏷️ Name synced by id: {cur.rowcount} rows")

        # 5) Backfill missing neighborhood_id by matching neighborhood name (case-insensitive)
        #    This fixes rows where id is NULL but name exists and is in the dictionary.
        cur.execute("""
            UPDATE berlin_source_data.theaters t
            SET neighborhood_id = d.neighborhood_id
            FROM tmp_neigh_dict d
            WHERE t.neighborhood_id IS NULL
              AND t.neighborhood IS NOT NULL
              AND LOWER(t.neighborhood) = LOWER(d.neighborhood);
        """)
        print(f"🧩 Backfilled id by name: {cur.rowcount} rows")

        # 6) After backfill, ensure all ids are 4 digits (safety pass)
        cur.execute("""
            UPDATE berlin_source_data.theaters
            SET neighborhood_id = LPAD(REGEXP_REPLACE(COALESCE(neighborhood_id,''), '\\D', '', 'g'), 4, '0')
            WHERE neighborhood_id IS NOT NULL
              AND neighborhood_id !~ '^[0-9]{4}$';
        """)
        print(f"🔒 Final normalize pass: {cur.rowcount} rows")

        # 7) Enforce on schema level (type + CHECK)
        cur.execute("""
            -- Fix column type to fixed width (keeps leading zeros)
            ALTER TABLE berlin_source_data.theaters
            ALTER COLUMN neighborhood_id TYPE CHAR(4)
            USING neighborhood_id;

            -- Ensure format is always 4 digits
            DO $$
            BEGIN
              IF NOT EXISTS (
                  SELECT 1 FROM pg_constraint WHERE conname = 'theaters_neighborhood_id_4digits_chk'
              ) THEN
                ALTER TABLE berlin_source_data.theaters
                ADD CONSTRAINT theaters_neighborhood_id_4digits_chk
                CHECK (neighborhood_id ~ '^[0-9]{4}$');
              END IF;
            END$$;
        """)
        print("🧱 DB constraints in place")

    print("✅ Committed")

    # 8) Quick QA
    with conn.cursor() as cur:
        cur.execute("""
            SELECT
              SUM(CASE WHEN neighborhood_id ~ '^[0-9]{4}$' THEN 1 ELSE 0 END) AS ok_4,
              SUM(CASE WHEN neighborhood_id IS NULL THEN 1 ELSE 0 END)       AS nulls,
              SUM(CASE WHEN neighborhood_id !~ '^[0-9]{4}$' AND neighborhood_id IS NOT NULL THEN 1 ELSE 0 END) AS bad
            FROM berlin_source_data.theaters;
        """)
        print("QA:", cur.fetchone())

        cur.execute("""
            SELECT theater_id, name, neighborhood_id, neighborhood
            FROM berlin_source_data.theaters
            WHERE neighborhood_id IS NULL OR neighborhood IS NULL
            ORDER BY name
            LIMIT 10;
        """)
        rows = cur.fetchall()
        if rows:
            print("⚠️ Still missing (sample):")
            for r in rows:
                print(r)
        else:
            print("🎉 No missing neighborhood_id or neighborhood in sample.")

finally:
    conn.close()
    print("🔒 Connection closed")


💾 Backup OK
📚 Dictionary ready: 51 ids
🔧 Normalized id length: 260 rows touched
🧽 '0000' -> NULL: 0 rows
🏷️ Name synced by id: 0 rows
🧩 Backfilled id by name: 0 rows
🔒 Final normalize pass: 0 rows
🧱 DB constraints in place
✅ Committed
QA: (282, 0, 0)
🎉 No missing neighborhood_id or neighborhood in sample.
🔒 Connection closed


# Data enrichment and normalization pipeline is now fully successful 🎯

| Step                                              | Meaning                                                                                      | Result                                                |
| ------------------------------------------------- | -------------------------------------------------------------------------------------------- | ----------------------------------------------------- |
| 💾 **Backup OK**                                  | A full safety copy of the table was made before any updates.                                 | ✅ Backup table `theaters_backup_neigh_final` created. |
| 📚 **Dictionary ready: 51 ids**                   | Reference CSV contained 51 unique neighborhood mappings (LOR Ortsteile).                | ✅ Reference data loaded successfully.                 |
| 🔧 **Normalized id length: 260 rows touched**     | 260 theaters had 3-digit or non-standard IDs that were fixed to 4-digit codes.               | ✅ All IDs standardized.                               |
| 🧽 **'0000' → NULL**                              | No placeholder “0000” codes found — nothing to clean.                                        | ✅ 0 rows.                                             |
| 🏷️ **Name synced by id: 0 rows**                 | All neighborhood names were already correct for their IDs.                                   | ✅ No changes needed.                                  |
| 🧩 **Backfilled id by name: 0 rows**              | No missing IDs were inferred from names — every record had both.                             | ✅ Data consistent.                                    |
| 🔒 **Final normalize pass: 0 rows**               | No residual invalid formats remained after the fix.                                          | ✅ Clean.                                              |
| 🧱 **DB constraints in place**                    | Added a `CHECK` constraint enforcing 4-digit numeric IDs (and set column type to `CHAR(4)`). | ✅ Structure locked and safe.                          |
| ✅ **Committed**                                   | All updates were applied to the database.                                                    | ✅ Transaction complete.                               |
| **QA: (282, 0, 0)**                               | 282 valid rows, 0 nulls, 0 bad formats.                                                      | ✅ Everything correct.                                 |
| 🎉 **No missing neighborhood_id or neighborhood** | Final verification sample found no missing values.                                           | ✅ Fully enriched.                                     |


In [33]:
import pandas as pd
import psycopg2

# --- Reconnect  ---
conn = psycopg2.connect(
    host="127.0.0.1",
    port=5433,
    user="marianna_gokova",
    password="6n2b8nw9IfmNdyYY",
    dbname="layereddb",
    sslmode="require"
)

try:
    # --- 1️⃣ Count unique neighborhood_id ---
    q_unique = """
        SELECT COUNT(DISTINCT neighborhood_id) AS uniq_ids
        FROM berlin_source_data.theaters;
    """
    uniq = pd.read_sql(q_unique, conn)
    print(f"🏙️ Unique neighborhoods in theaters table: {int(uniq['uniq_ids'].iloc[0])}")

    # --- 2️⃣ Distribution by neighborhood (top 20) ---
    q_dist = """
        SELECT neighborhood_id, neighborhood, COUNT(*) AS n
        FROM berlin_source_data.theaters
        GROUP BY neighborhood_id, neighborhood
        ORDER BY n DESC
        LIMIT 20;
    """
    df_dist = pd.read_sql(q_dist, conn)
    print("\n📊 Top 20 neighborhoods by theater count:")
    display(df_dist)

    # --- 3️⃣ District ↔ neighborhood cross-check ---
    q_cross = """
        SELECT district_id, neighborhood_id, COUNT(*) AS n
        FROM berlin_source_data.theaters
        GROUP BY district_id, neighborhood_id
        ORDER BY district_id, neighborhood_id;
    """
    df_cross = pd.read_sql(q_cross, conn)
    print("\n🧩 District ↔ Neighborhood mapping (sample):")
    display(df_cross.head(20))

finally:
    conn.close()
    print("\n🔒 Connection closed")


🏙️ Unique neighborhoods in theaters table: 51

📊 Top 20 neighborhoods by theater count:


/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/2127692855.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  uniq = pd.read_sql(q_unique, conn)
/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/2127692855.py:31: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_dist = pd.read_sql(q_dist, conn)


neighborhood_id     neighborhood   n
0             0101            Mitte  36
1             0202        Kreuzberg  27
2             0401   Charlottenburg  25
3             0301  Prenzlauer Berg  20
4             0801         Neukölln  17
5             0201   Friedrichshain  17
6             0104       Tiergarten  10
7             0701       Schöneberg  10
8             0402      Wilmersdorf   9
9             0703        Tempelhof   7
10            0501          Spandau   6
11            0102           Moabit   6
12            0105          Wedding   6
13            0601         Steglitz   6
14            0911  Friedrichshagen   5
15            0702        Friedenau   5
16            0302        Weißensee   5
17            0910         Köpenick   5
18            0909  Oberschöneweide   4
19            1001          Marzahn   4


🧩 District ↔ Neighborhood mapping (sample):


/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_43442/2127692855.py:42: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_cross = pd.read_sql(q_cross, conn)


district_id neighborhood_id   n
0     11001001            0101  36
1     11001001            0102   6
2     11001001            0103   1
3     11001001            0104  10
4     11001001            0105   6
5     11001001            0106   4
6     11002002            0201  17
7     11002002            0202  27
8     11003003            0301  20
9     11003003            0302   5
10    11003003            0307   3
11    11003003            0311   1
12    11004004            0401  25
13    11004004            0402   9
14    11004004            0404   1
15    11004004            0405   2
16    11004004            0407   1
17    11005005            0501   6
18    11005005            0502   1
19    11006006            0601   6


🔒 Connection closed
